File: rs/src/indicators/common/rate_of_change/rate_of_change.rs
use crate::entities::bar::Bar;
use crate::entities::bar_component::{component_value as bar_component_value, BarComponent, DEFAULT_BAR_COMPONENT};
use crate::entities::quote::Quote;
use crate::entities::quote_component::{component_value as quote_component_value, QuoteComponent, DEFAULT_QUOTE_COMPONENT};
use crate::entities::scalar::Scalar;
use crate::entities::trade::Trade;
use crate::entities::trade_component::{component_value as trade_component_value, TradeComponent, DEFAULT_TRADE_COMPONENT};
use crate::indicators::core::build_metadata::{build_metadata, OutputText};
use crate::indicators::core::component_triple_mnemonic::component_triple_mnemonic;
use crate::indicators::core::identifier::Identifier;
use crate::indicators::core::indicator::{Indicator, Output};
use crate::indicators::core::line_indicator::LineIndicator;
use crate::indicators::core::metadata::Metadata;

// ---------------------------------------------------------------------------
// Params
// ---------------------------------------------------------------------------

/// Parameters to create an instance of the rate of change indicator.
pub struct RateOfChangeParams {
    /// The length (number of time periods).
    /// Must be greater than 0.
    pub length: usize,
    /// Bar component to extract. `None` means use default (Close).
    pub bar_component: Option<BarComponent>,
    /// Quote component to extract. `None` means use default (Mid).
    pub quote_component: Option<QuoteComponent>,
    /// Trade component to extract. `None` means use default (Price).
    pub trade_component: Option<TradeComponent>,
}

impl Default for RateOfChangeParams {
    fn default() -> Self {
        Self {
            length: 10,
            bar_component: None,
            quote_component: None,
            trade_component: None,
        }
    }
}

// ---------------------------------------------------------------------------
// Output
// ---------------------------------------------------------------------------

/// Enumerates the outputs of the rate of change indicator.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum RateOfChangeOutput {
    /// The scalar value of the rate of change.
    Value = 1,
}

// ---------------------------------------------------------------------------
// Indicator
// ---------------------------------------------------------------------------

/// Computes the Rate of Change (ROC).
///
/// ROCi = (Pi - Pi-l) / Pi-l * 100
///
/// where l is the length.
///
/// The indicator is not primed during the first l updates.
pub struct RateOfChange {
    line: LineIndicator,
    window: Vec<f64>,
    window_length: usize,
    window_count: usize,
    last_index: usize,
    primed: bool,
}

impl RateOfChange {
    /// Creates a new RateOfChange from the given parameters.
    pub fn new(params: &RateOfChangeParams) -> Result<Self, String> {
        if params.length < 1 {
            return Err("invalid rate of change parameters: length should be positive".to_string());
        }

        let bc = params.bar_component.unwrap_or(DEFAULT_BAR_COMPONENT);
        let qc = params.quote_component.unwrap_or(DEFAULT_QUOTE_COMPONENT);
        let tc = params.trade_component.unwrap_or(DEFAULT_TRADE_COMPONENT);

        let bar_func = bar_component_value(bc);
        let quote_func = quote_component_value(qc);
        let trade_func = trade_component_value(tc);

        let mnemonic = format!("roc({}{})", params.length, component_triple_mnemonic(bc, qc, tc));
        let description = format!("Rate of Change {}", mnemonic);

        let line = LineIndicator::new(mnemonic, description, bar_func, quote_func, trade_func);

        let window_length = params.length + 1;

        Ok(Self {
            line,
            window: vec![0.0; window_length],
            window_length,
            window_count: 0,
            last_index: params.length,
            primed: false,
        })
    }

    /// Core update logic. Returns the rate of change value or NaN if not yet primed.
    pub fn update(&mut self, sample: f64) -> f64 {
        if sample.is_nan() {
            return sample;
        }

        const EPSILON: f64 = 1e-13;

        if self.primed {
            for i in 0..self.last_index {
                self.window[i] = self.window[i + 1];
            }

            self.window[self.last_index] = sample;
            let previous = self.window[0];

            if previous.abs() > EPSILON {
                return (sample / previous - 1.0) * 100.0;
            }

            return 0.0;
        }

        self.window[self.window_count] = sample;
        self.window_count += 1;

        if self.window_length == self.window_count {
            self.primed = true;
            let previous = self.window[0];

            if previous.abs() > EPSILON {
                return (sample / previous - 1.0) * 100.0;
            }

            return 0.0;
        }

        f64::NAN
    }
}

impl Indicator for RateOfChange {
    fn is_primed(&self) -> bool {
        self.primed
    }

    fn metadata(&self) -> Metadata {
        build_metadata(
            Identifier::RateOfChange,
            &self.line.mnemonic,
            &self.line.description,
            &[OutputText {
                mnemonic: self.line.mnemonic.clone(),
                description: self.line.description.clone(),
            }],
        )
    }

    fn update_scalar(&mut self, sample: &Scalar) -> Output {
        let value = self.update(sample.value);
        vec![Box::new(Scalar::new(sample.time, value))]
    }

    fn update_bar(&mut self, sample: &Bar) -> Output {
        let sample_value = (self.line.bar_func)(sample);
        let value = self.update(sample_value);
        vec![Box::new(Scalar::new(sample.time, value))]
    }

    fn update_quote(&mut self, sample: &Quote) -> Output {
        let sample_value = (self.line.quote_func)(sample);
        let value = self.update(sample_value);
        vec![Box::new(Scalar::new(sample.time, value))]
    }

    fn update_trade(&mut self, sample: &Trade) -> Output {
        let sample_value = (self.line.trade_func)(sample);
        let value = self.update(sample_value);
        vec![Box::new(Scalar::new(sample.time, value))]
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use super::super::testdata::testdata;
    use crate::entities::bar_component::BarComponent;
    use crate::entities::quote_component::QuoteComponent;
    use crate::entities::trade_component::TradeComponent;
    use crate::indicators::core::outputs::shape::Shape;
    fn create_roc(length: usize) -> RateOfChange {
        RateOfChange::new(&RateOfChangeParams { length, ..Default::default() }).unwrap()
    }

    #[test]
    fn test_update_length_14() {
        let mut roc = create_roc(14);
        let input = testdata::test_input();

        for i in 0..13 {
            assert!(roc.update(input[i]).is_nan(), "[{}] expected NaN", i);
        }

        for i in 13..input.len() {
            let act = roc.update(input[i]);

            match i {
                14 => assert!((act - (-0.546)).abs() < 1e-2, "[14] expected -0.546, got {}", act),
                15 => assert!((act - (-2.109)).abs() < 1e-2, "[15] expected -2.109, got {}", act),
                16 => assert!((act - (-5.53)).abs() < 1e-2, "[16] expected -5.53, got {}", act),
                251 => assert!((act - (-1.0367)).abs() < 1e-2, "[251] expected -1.0367, got {}", act),
                _ => {}
            }
        }

        assert!(roc.update(f64::NAN).is_nan());
    }

    #[test]
    fn test_update_entity() {
        let length = 2;
        let inp = 3.0_f64;
        let exp = 0.0_f64; // roc = (3-3)/3 * 100 = 0
        let time = 1617235200;

        // scalar
        let mut roc = create_roc(length);
        roc.update(inp);
        roc.update(inp);
        let out = roc.update_scalar(&Scalar::new(time, inp));
        assert_eq!(out.len(), 1);
        let s = out[0].downcast_ref::<Scalar>().unwrap();
        assert_eq!(s.time, time);
        assert!((s.value - exp).abs() < 1e-13);

        // bar
        let mut roc = create_roc(length);
        roc.update(inp);
        roc.update(inp);
        let bar = Bar::new(time, 0.0, 0.0, 0.0, inp, 0.0);
        let out = roc.update_bar(&bar);
        let s = out[0].downcast_ref::<Scalar>().unwrap();
        assert!((s.value - exp).abs() < 1e-13);

        // quote
        let mut roc = create_roc(length);
        roc.update(inp);
        roc.update(inp);
        let quote = Quote::new(time, inp, inp, 0.0, 0.0);
        let out = roc.update_quote(&quote);
        let s = out[0].downcast_ref::<Scalar>().unwrap();
        assert!((s.value - exp).abs() < 1e-13);

        // trade
        let mut roc = create_roc(length);
        roc.update(inp);
        roc.update(inp);
        let trade = Trade::new(time, inp, 0.0);
        let out = roc.update_trade(&trade);
        let s = out[0].downcast_ref::<Scalar>().unwrap();
        assert!((s.value - exp).abs() < 1e-13);
    }

    #[test]
    fn test_is_primed() {
        let input = testdata::test_input();

        for &length in &[1_usize, 2, 3, 5, 10] {
            let mut roc = create_roc(length);
            assert!(!roc.is_primed());

            for i in 0..length {
                roc.update(input[i]);
                assert!(!roc.is_primed(), "length={}, [{}] should not be primed", length, i);
            }

            for i in length..input.len() {
                roc.update(input[i]);
                assert!(roc.is_primed(), "length={}, [{}] should be primed", length, i);
            }
        }
    }

    #[test]
    fn test_metadata() {
        let roc = create_roc(5);
        let m = roc.metadata();
        assert_eq!(m.identifier, Identifier::RateOfChange);
        assert_eq!(m.outputs.len(), 1);
        assert_eq!(m.outputs[0].kind, RateOfChangeOutput::Value as i32);
        assert_eq!(m.outputs[0].shape, Shape::Scalar);
        assert_eq!(m.outputs[0].mnemonic, "roc(5)");
        assert_eq!(m.outputs[0].description, "Rate of Change roc(5)");
    }

    #[test]
    fn test_new_invalid_length() {
        let r = RateOfChange::new(&RateOfChangeParams { length: 0, ..Default::default() });
        assert!(r.is_err());
        assert_eq!(r.err().unwrap(), "invalid rate of change parameters: length should be positive");
    }

    #[test]
    fn test_mnemonic_components() {
        let roc = create_roc(5);
        assert_eq!(roc.line.mnemonic, "roc(5)");

        let roc = RateOfChange::new(&RateOfChangeParams {
            length: 5, bar_component: Some(BarComponent::Median), ..Default::default()
        }).unwrap();
        assert_eq!(roc.line.mnemonic, "roc(5, hl/2)");

        let roc = RateOfChange::new(&RateOfChangeParams {
            length: 5, quote_component: Some(QuoteComponent::Bid), ..Default::default()
        }).unwrap();
        assert_eq!(roc.line.mnemonic, "roc(5, b)");

        let roc = RateOfChange::new(&RateOfChangeParams {
            length: 5, trade_component: Some(TradeComponent::Volume), ..Default::default()
        }).unwrap();
        assert_eq!(roc.line.mnemonic, "roc(5, v)");

        let roc = RateOfChange::new(&RateOfChangeParams {
            length: 5,
            bar_component: Some(BarComponent::High),
            quote_component: None,
            trade_component: Some(TradeComponent::Volume),
        }).unwrap();
        assert_eq!(roc.line.mnemonic, "roc(5, h, v)");
    }
}

File: zig/src/indicators/common/rate_of_change/rate_of_change.zig
const std = @import("std");
const math = std.math;


const entities = @import("entities");
const Bar = entities.Bar;
const Quote = entities.Quote;
const Trade = entities.Trade;
const Scalar = entities.Scalar;
const bar_component = entities.bar_component;
const quote_component = entities.quote_component;
const trade_component = entities.trade_component;
const indicator_mod = @import("../../core/indicator.zig");
const line_indicator_mod = @import("../../core/line_indicator.zig");
const build_metadata_mod = @import("../../core/build_metadata.zig");
const component_triple_mnemonic_mod = @import("../../core/component_triple_mnemonic.zig");
const identifier_mod = @import("../../core/identifier.zig");
const metadata_mod = @import("../../core/metadata.zig");

const OutputArray = indicator_mod.OutputArray;
const LineIndicator = line_indicator_mod.LineIndicator;
const Identifier = identifier_mod.Identifier;
const Metadata = metadata_mod.Metadata;

/// Enumerates the outputs of the rate of change indicator.
pub const RateOfChangeOutput = enum(u8) {
    value = 1,
};

/// Parameters to create an instance of the rate of change indicator.
pub const RateOfChangeParams = struct {
    length: usize,
    bar_component: ?bar_component.BarComponent = null,
    quote_component: ?quote_component.QuoteComponent = null,
    trade_component: ?trade_component.TradeComponent = null,
};

/// Computes the Rate of Change (ROC).
///
/// ROCi = 100 * (Pi / Pi-l - 1)
///
/// The indicator is not primed during the first l updates.
pub const RateOfChange = struct {
    line: LineIndicator,
    window: []f64,
    window_length: usize,
    window_count: usize,
    last_index: usize,
    primed: bool,
    allocator: std.mem.Allocator,
    mnemonic_buf: [64]u8,
    mnemonic_len: usize,
    description_buf: [128]u8,
    description_len: usize,

    const epsilon = 1e-13;

    pub fn init(allocator: std.mem.Allocator, params: RateOfChangeParams) !RateOfChange {
        if (params.length < 1) {
            return error.InvalidLength;
        }

        const bc = params.bar_component orelse bar_component.default_bar_component;
        const qc = params.quote_component orelse quote_component.default_quote_component;
        const tc = params.trade_component orelse trade_component.default_trade_component;

        var triple_buf: [64]u8 = undefined;
        const triple = component_triple_mnemonic_mod.componentTripleMnemonic(&triple_buf, bc, qc, tc);

        var mnemonic_buf: [64]u8 = undefined;
        const mnemonic_slice = std.fmt.bufPrint(&mnemonic_buf, "roc({d}{s})", .{ params.length, triple }) catch
            return error.MnemonicTooLong;
        const mnemonic_len = mnemonic_slice.len;

        var description_buf: [128]u8 = undefined;
        const desc_slice = std.fmt.bufPrint(&description_buf, "Rate of Change {s}", .{mnemonic_slice}) catch
            return error.MnemonicTooLong;
        const description_len = desc_slice.len;

        const window_length = params.length + 1;
        const window = try allocator.alloc(f64, window_length);
        @memset(window, 0.0);

        return .{
            .line = LineIndicator.new(
                mnemonic_buf[0..mnemonic_len],
                description_buf[0..description_len],
                params.bar_component,
                params.quote_component,
                params.trade_component,
            ),
            .window = window,
            .window_length = window_length,
            .window_count = 0,
            .last_index = params.length,
            .primed = false,
            .allocator = allocator,
            .mnemonic_buf = mnemonic_buf,
            .mnemonic_len = mnemonic_len,
            .description_buf = description_buf,
            .description_len = description_len,
        };
    }

    pub fn deinit(self: *RateOfChange) void {
        self.allocator.free(self.window);
    }

    pub fn fixSlices(self: *RateOfChange) void {
        self.line.mnemonic = self.mnemonic_buf[0..self.mnemonic_len];
        self.line.description = self.description_buf[0..self.description_len];
    }

    fn compute(sample: f64, previous: f64) f64 {
        if (@abs(previous) > epsilon) {
            return (sample / previous - 1.0) * 100.0;
        }
        return 0.0;
    }

    pub fn update(self: *RateOfChange, sample: f64) f64 {
        if (math.isNan(sample)) {
            return sample;
        }

        if (self.primed) {
            if (self.last_index > 1) {
                var i: usize = 0;
                while (i < self.last_index) : (i += 1) {
                    self.window[i] = self.window[i + 1];
                }
            }

            self.window[self.last_index] = sample;
            return compute(sample, self.window[0]);
        }

        self.window[self.window_count] = sample;
        self.window_count += 1;

        if (self.window_length == self.window_count) {
            self.primed = true;
            return compute(sample, self.window[0]);
        }

        return math.nan(f64);
    }

    pub fn isPrimed(self: *const RateOfChange) bool {
        return self.primed;
    }

    pub fn getMetadata(self: *const RateOfChange, out: *Metadata) void {
        build_metadata_mod.buildMetadata(
            out,
            .rate_of_change,
            self.line.mnemonic,
            self.line.description,
            &[_]build_metadata_mod.OutputText{
                .{ .mnemonic = self.line.mnemonic, .description = self.line.description },
            },
        );
    }

    pub fn updateScalar(self: *RateOfChange, sample: *const Scalar) OutputArray {
        const value = self.update(sample.value);
        return LineIndicator.wrapScalar(sample.time, value);
    }

    pub fn updateBar(self: *RateOfChange, sample: *const Bar) OutputArray {
        const value = self.update(self.line.extractBar(sample));
        return LineIndicator.wrapScalar(sample.time, value);
    }

    pub fn updateQuote(self: *RateOfChange, sample: *const Quote) OutputArray {
        const value = self.update(self.line.extractQuote(sample));
        return LineIndicator.wrapScalar(sample.time, value);
    }

    pub fn updateTrade(self: *RateOfChange, sample: *const Trade) OutputArray {
        const value = self.update(self.line.extractTrade(sample));
        return LineIndicator.wrapScalar(sample.time, value);
    }

    pub fn indicator(self: *RateOfChange) indicator_mod.Indicator {
        return .{ .ptr = @ptrCast(self), .vtable = &vtable };
    }

    const vtable = indicator_mod.Indicator.VTable{
        .isPrimed = vtableIsPrimed,
        .metadata = vtableMetadata,
        .updateScalar = vtableUpdateScalar,
        .updateBar = vtableUpdateBar,
        .updateQuote = vtableUpdateQuote,
        .updateTrade = vtableUpdateTrade,
    };

    fn vtableIsPrimed(ptr: *anyopaque) bool {
        const self: *RateOfChange = @ptrCast(@alignCast(ptr));
        return self.isPrimed();
    }
    fn vtableMetadata(ptr: *anyopaque, out: *Metadata) void {
        const self: *const RateOfChange = @ptrCast(@alignCast(ptr));
        self.getMetadata(out);
    }
    fn vtableUpdateScalar(ptr: *anyopaque, sample: *const Scalar) OutputArray {
        const self: *RateOfChange = @ptrCast(@alignCast(ptr));
        return self.updateScalar(sample);
    }
    fn vtableUpdateBar(ptr: *anyopaque, sample: *const Bar) OutputArray {
        const self: *RateOfChange = @ptrCast(@alignCast(ptr));
        return self.updateBar(sample);
    }
    fn vtableUpdateQuote(ptr: *anyopaque, sample: *const Quote) OutputArray {
        const self: *RateOfChange = @ptrCast(@alignCast(ptr));
        return self.updateQuote(sample);
    }
    fn vtableUpdateTrade(ptr: *anyopaque, sample: *const Trade) OutputArray {
        const self: *RateOfChange = @ptrCast(@alignCast(ptr));
        return self.updateTrade(sample);
    }

    pub const InitError = error{ InvalidLength, MnemonicTooLong, OutOfMemory };
};

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

const testing = std.testing;
const testdata = @import("testdata.zig");


fn createRoc(allocator: std.mem.Allocator, length: usize) !RateOfChange {
    var roc = try RateOfChange.init(allocator, .{ .length = length });
    roc.fixSlices();
    return roc;
}

test "roc update length 14" {
    const input = testdata.testInput();
    var roc = try createRoc(testing.allocator, 14);
    defer roc.deinit();

    for (0..13) |i| {
        try testing.expect(math.isNan(roc.update(input[i])));
    }

    for (13..252) |i| {
        const act = roc.update(input[i]);
        if (i == 14) try testing.expect(@abs(act - (-0.546)) < 1e-2);
        if (i == 15) try testing.expect(@abs(act - (-2.109)) < 1e-2);
        if (i == 16) try testing.expect(@abs(act - (-5.53)) < 1e-2);
        if (i == 251) try testing.expect(@abs(act - (-1.0367)) < 1e-2);
    }

    try testing.expect(math.isNan(roc.update(math.nan(f64))));
}

test "roc is primed" {
    const input = testdata.testInput();
    inline for ([_]usize{ 1, 2, 5, 10 }) |length| {
        var roc = try createRoc(testing.allocator, length);
        defer roc.deinit();
        try testing.expect(!roc.isPrimed());
        for (0..length) |i| {
            _ = roc.update(input[i]);
            try testing.expect(!roc.isPrimed());
        }
        for (length..252) |i| {
            _ = roc.update(input[i]);
            try testing.expect(roc.isPrimed());
        }
    }
}

test "roc metadata" {
    var roc = try createRoc(testing.allocator, 5);
    defer roc.deinit();
    var m: Metadata = undefined;
    roc.getMetadata(&m);
    try testing.expectEqual(Identifier.rate_of_change, m.identifier);
    try testing.expectEqual(@as(usize, 1), m.outputs_len);
    try testing.expectEqualStrings("roc(5)", m.outputs_buf[0].mnemonic);
    try testing.expectEqualStrings("Rate of Change roc(5)", m.outputs_buf[0].description);
}

test "roc init invalid length" {
    try testing.expectError(error.InvalidLength, RateOfChange.init(testing.allocator, .{ .length = 0 }));
}

test "roc update entity" {
    const length: usize = 2;
    const inp: f64 = 3.0;
    const exp: f64 = 0.0; // (3/3 - 1) * 100 = 0
    const time: i64 = 1617235200;

    // scalar
    {
        var roc = try createRoc(testing.allocator, length);
        defer roc.deinit();
        _ = roc.update(inp);
        _ = roc.update(inp);
        const out = roc.updateScalar(&.{ .time = time, .value = inp });
        try testing.expectEqual(@as(usize, 1), out.len);
        const s = out.slice()[0].scalar;
        try testing.expectEqual(time, s.time);
        try testing.expect(@abs(s.value - exp) < 1e-13);
    }
}
